x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java

```diff
@@ -1249,60 +1255,137 @@ public String getFeatureDescription() {
             // Stop data frame analytics
-            StopDataFrameAnalyticsAction.Request  stopDataFramesReq = new StopDataFrameAnalyticsAction.Request("_all");
-            stopDataFramesReq.setForce(true);
-            stopDataFramesReq.setAllowNoMatch(true);
+            StopDataFrameAnalyticsAction.Request  stopDataFramesReq = new StopDataFrameAnalyticsAction.Request("_all")
+                .setForce(true)
+                .setAllowNoMatch(true);
             client.execute(StopDataFrameAnalyticsAction.INSTANCE, stopDataFramesReq, afterDataframesStopped);
-        }, listener::onFailure);
+        }, unsetResetModeListener::onFailure);
 
         // Close anomaly detection jobs
         ActionListener<StopDatafeedAction.Response> afterDataFeedsStopped = ActionListener.wrap(datafeedResponse -> {
             // Handle the response
             results.put("datafeeds", datafeedResponse.isStopped());
 
             // Close anomaly detection jobs
-            CloseJobAction.Request closeJobsRequest = new CloseJobAction.Request();
-            closeJobsRequest.setForce(true);
-            closeJobsRequest.setAllowNoMatch(true);
-            closeJobsRequest.setJobId("_all");
+            CloseJobAction.Request closeJobsRequest = new CloseJobAction.Request()
+                .setForce(true)
+                .setAllowNoMatch(true)
+                .setJobId("_all");
             client.execute(CloseJobAction.INSTANCE, closeJobsRequest, afterAnomalyDetectionClosed);
-        }, listener::onFailure);
+        }, unsetResetModeListener::onFailure);
 
         // Stop data feeds
-        StopDatafeedAction.Request stopDatafeedsReq = new StopDatafeedAction.Request("_all");
-        stopDatafeedsReq.setAllowNoMatch(true);
-        stopDatafeedsReq.setForce(true);
-        client.execute(StopDatafeedAction.INSTANCE, stopDatafeedsReq,
-            afterDataFeedsStopped);
+        ActionListener<AcknowledgedResponse> pipelineValidation = ActionListener.wrap(
+            acknowledgedResponse -> {
+                StopDatafeedAction.Request stopDatafeedsReq = new StopDatafeedAction.Request("_all")
+                    .setAllowNoMatch(true)
+                    .setForce(true);
+                client.execute(StopDatafeedAction.INSTANCE, stopDatafeedsReq,
+                    afterDataFeedsStopped);
+            },
+            unsetResetModeListener::onFailure
+        );
+
+        // validate no pipelines are using machine learning models
+        ActionListener<AcknowledgedResponse> afterResetModeSet = ActionListener.wrap(
+            acknowledgedResponse -> {
+                int numberInferenceProcessors = countNumberInferenceProcessors(clusterService.state());
+                if (numberInferenceProcessors > 0) {
+                    unsetResetModeListener.onFailure(
+                        new RuntimeException(
+                            "Unable to reset component as there are ingest pipelines still referencing trained machine learning models"
```

Comment:
Does this mean we're still going to need custom cleanup in between integration tests?  If so it devalues this API a lot.

I would say that if we blow away ML entirely we blow away the ingest pipelines that are referencing ML.  To paraphrase Tony Blair: tough on ML, tough on the _users_ of ML.

Reply:
@droberts195 that feels very rough to me. 

All other ML interactions are sort of their own thing. If we go and delete ingest pipelines, that could blow up a users cluster. Maybe they don't honestly know if a pipeline is in use.

I think safety here is better than causing a user to potentially lose data.

Also, we already have this custom cleanup code in the tests that use pipelines (we have this outside of the normal clean up code).

Reply:
> Also, we already have this custom cleanup code in the tests that use pipelines (we have this outside of the normal clean up code).

OK, in that case this PR can stay as-is and we can decide what to do about ingest pipelines separately.

Once this PR is merged we should be able to dramatically simplify our test cleanup and see if anything breaks as a result (which would reveal more things our reset needs to do).